File: controls.py
from worldgen import getBlockCoord, getChunkCoord, generateChunkTerrain
from globalVariables import chunks, keys, keysPressed
import pygame


class Mouse():
    def __init__(self):
        self.x = 0
        self.y = 0
        self.pos = (0, 0)
        self.cameraRelativeX = 0
        self.cameraRelativeZ = 0
        self.cameraRelativePos = (0, 0)
        # block height for mining / placing
        self.selectedY = 0
        self.selectedYChange = 0
        # detection for blocks and stuff
        self.hoveredBlock = {}
        self.hoveredSlotId = 0
        

        self.inPlayerInventory = False
        self.inPlayerHotbar = False
        self.inASlot = False
        # needs to have these two in order to tranfer item data properly
        self.heldItem = {
            "contents": "empty",
            "count": 0
        }

        self.buttons = {
            "left": False,
            "middle": False,
            "right": False,
            "pressed": {
                "left": False,
                "middle": False,
                "right": False
            },
            "held": {
                "left": False,
                "middle": False,
                "right": False
            }
        }

mouse = Mouse()


def updateMouseAndKeys():
    
    tempKeys = pygame.key.get_pressed()
        
    for keyID, thing in enumerate(tempKeys):
        keysPressed[keyID] = False
        if not keys[0][keyID] and tempKeys[keyID]:
            keysPressed[keyID] = True
    keys[0] = tempKeys

    mouse.pos = pygame.mouse.get_pos()
    mouse.x, mouse.y = mouse.pos[0], mouse.pos[1]

    mouse.buttons["pressed"]["left"] = False
    mouse.buttons["pressed"]["middle"] = False
    mouse.buttons["pressed"]["right"] = False


    mouseButtons = pygame.mouse.get_pressed()

    if not mouse.buttons["left"]:
        if mouseButtons[0]:
            mouse.buttons["pressed"]["left"] = True

    if not mouse.buttons["middle"]:
        if mouseButtons[1]:
            mouse.buttons["pressed"]["middle"] = True

    if not mouse.buttons["right"]:
        if mouseButtons[2]:
            mouse.buttons["pressed"]["right"] = True

    mouse.buttons["left"] = mouseButtons[0]
    mouse.buttons["middle"] = mouseButtons[1]
    mouse.buttons["right"] = mouseButtons[2]

    x = mouse.cameraRelativeX
    y = mouse.selectedY
    z = mouse.cameraRelativeZ
    
    chunkCoord = getChunkCoord(x, z)
    blockCoord = getBlockCoord(x, y, z)
    try:
        chunks[chunkCoord]["data"][blockCoord]
    except:
        generateChunkTerrain(chunkCoord)
    mouse.hoveredBlock["block"] = chunks[chunkCoord]["data"][blockCoord]
    mouse.hoveredBlock["chunkCoord"] = chunkCoord
    mouse.hoveredBlock["blockCoord"] = blockCoord
    
    






    

File: Controls.mjs
import { chunks, keys, keysPressed, showLoadingProgress, canvas, Rect, mouse } from "./GlobalVariables.mjs";
showLoadingProgress("loading Controls.mjs");
import { getBlockCoord, getChunkCoord, generateChunkTerrain } from "./Worldgen.mjs";

let inFocus = false;
let commandInputFocused = false;

let commandInput = document.getElementById("command input");
commandInput.onfocus = () => {
    commandInputFocused = true;
}
commandInput.onblur = () => {
    commandInputFocused = false;
}

canvas.addEventListener("mouseenter", function (event) {
    inFocus = true;
})
canvas.addEventListener("mouseleave", function (event) {
    inFocus = false;
})

canvas.addEventListener("mousemove", function (event) {
    const canvasRect = canvas.getBoundingClientRect();
    const x = event.clientX - canvasRect.left;
    const y = event.clientY - canvasRect.top;
    mouse.pos = [x, y];
    mouse.x = x;
    mouse.y = y;
});


addEventListener("contextmenu", function (event) {
    if (inFocus) { // disable the context menu while playing game
        event.preventDefault();
    };
})

window.addEventListener("mousedown", function (event) {

    if (inFocus && !commandInputFocused) {

        event.preventDefault();

        switch (event.button) {
            case 0:
                mouse.buttons.left = true;
                mouse.buttons.pressed.left = true;
                break;
            case 2:
                mouse.buttons.right = true;
                mouse.buttons.pressed.right = true;
                break;
        };
    };
});

window.addEventListener("mouseup", function (event) {

    if (inFocus) {
        switch (event.button) {
            case 0:
                mouse.buttons.left = false; break;
            case 2:
                mouse.buttons.right = false; break;
        }
    };

});



window.addEventListener("keydown", function (event) {

    if (inFocus && !commandInputFocused) {
        event.preventDefault();

        if (!keys[event.key]) { keysPressed[event.key] = true; };
        keys[event.key] = true;
        keys.ctrl = event.ctrlKey;
        keys.shift = event.shiftKey;

    };

})

window.addEventListener("keyup", function (event) {

    if (inFocus && !commandInputFocused) {
        keys[event.key] = false;
        keys.ctrl = event.ctrlKey;
        keys.shift = event.shiftKey;

        if (event.key == "/") {
            document.getElementById("command input").focus();
        }
    };

})
// add an event listener for clicking, keys down, keys up, and maybe some others?

export function updateMouseAndKeys() {

    let x = mouse.cameraRelativeX;
    let y = mouse.selectedY;
    let z = mouse.cameraRelativeZ;

    let chunkCoord = getChunkCoord(x, z);
    let blockCoord = getBlockCoord(x, y, z);

    if (chunks[chunkCoord].data[blockCoord] === undefined) { generateChunkTerrain(chunkCoord); };

    mouse.hoveredBlock = chunks[chunkCoord].data[blockCoord];
    mouse.hoveredBlock.chunkCoord = chunkCoord;
    mouse.hoveredBlock.blockCoord = blockCoord;


    mouse.buttons.pressed.left = false;
    mouse.buttons.pressed.middle = false;
    mouse.buttons.pressed.right = false;
    for (const key of Object.keys(keysPressed)) {
        keysPressed[key] = false;
    }

}

showLoadingProgress("controls initialized");

